File: Java_Version/8_String_to_Integer_(atoi).java
/*Implement atoi to convert a string to an integer.

Hint: Carefully consider all possible input cases. If you want a challenge,
please do not see below and ask yourself what are the possible input cases.

Notes: It is intended for this problem to be specified vaguely (ie, no given
input specs). You are responsible to gather all the input requirements up
front.


Requirements for atoi:

  The function first discards as many whitespace characters as necessary
until the first non-whitespace character is found. Then, starting from this
character, takes an optional initial plus or minus sign followed by as many
numerical digits as possible, and interprets them as a numerical value.

The string can contain additional characters after those that form the
integral number, which are ignored and have no effect on the behavior of this
function.

If the first sequence of non-whitespace characters in str is not a valid
integral number, or if no such sequence exists because either str is empty or
it contains only whitespace characters, no conversion is performed.

If no valid conversion could be performed, a zero value is returned. If the
correct value is out of the range of representable values, INT_MAX (2147483647)
or INT_MIN (-2147483648) is returned.*/



class Solution {

    public int myAtoi(String str) {

        if (str == null || str.length() == 0) {
            return 0;
        }

        int currentIndex = 0;
        while (currentIndex < str.length()
                && str.charAt(currentIndex) == ' ') {
            currentIndex++;
        }

        if (currentIndex == str.length()) {
            return 0;
        }


        boolean isPositive = true;
        if (str.charAt(currentIndex) == '+') {
            currentIndex++;
        } else if (str.charAt(currentIndex) == '-') {
            isPositive = false;
            currentIndex++;
        }

        return isPositive ?
                handlePositiveInteger(str, currentIndex) :
                handleNegativeInteger(str, currentIndex);
    }

    private int handlePositiveInteger(
            String str, int startIndex) {

        long result = 0;

        while (startIndex < str.length()) {

            char currentChar = str.charAt(startIndex);

            if (currentChar >= '0' && currentChar <= '9') {

                result = result * 10 + (currentChar - '0');

                if (result > Integer.MAX_VALUE) {
                    return Integer.MAX_VALUE;
                }

            } else {
                break;
            }

            startIndex++;
        }

        return (int) result;
    }

    private int handleNegativeInteger(
            String str, int startIndex) {

        long result = 0;

        while (startIndex < str.length()) {

            char currentChar = str.charAt(startIndex);

            if (currentChar >= '0' && currentChar <= '9') {

                result = result * 10 - (currentChar - '0');

                if (result < Integer.MIN_VALUE) {
                    return Integer.MIN_VALUE;
                }

            } else {
                break;
            }

            startIndex++;
        }

        return (int) result;
    }
}

File: C++_Version/8_String_to_Integer_(atoi).cpp
/*
  Implement atoi which converts a string to an integer.

  The function first discards as many whitespace characters as necessary until
the first non-whitespace character is found. Then, starting from this character,
takes an optional initial plus or minus sign followed by as many numerical
digits as possible, and interprets them as a numerical value.

  The string can contain additional characters after those that form the
integral number, which are ignored and have no effect on the behavior of this
function.

  If the first sequence of non-whitespace characters in str is not a valid
integral number, or if no such sequence exists because either str is empty or
it contains only whitespace characters, no conversion is performed.

  If no valid conversion could be performed, a zero value is returned.

Note:

  1. Only the space character ' ' is considered as whitespace character.

  2. Assume we are dealing with an environment which could only store integers
  within the 32-bit signed integer range: [−2^31,  2^31 − 1]. If the numerical
  value is out of the range of representable values, INT_MAX (2^31 − 1) or
  INT_MIN (−2^31) is returned.

Example 1:

  Input: "42"
  Output: 42

Example 2:

  Input: "   -42"
  Output: -42

  Explanation: The first non-whitespace character is '-', which is the minus
  sign. Then take as many numerical digits as possible, which gets 42.

Example 3:

  Input: "4193 with words"
  Output: 4193

  Explanation: Conversion stops at digit '3' as the next character is not a
  numerical digit.

Example 4:

  Input: "words and 987"
  Output: 0

  Explanation: The first non-whitespace character is 'w', which is not a
  numerical digit or a +/- sign. Therefore no valid conversion could be
  performed.

Example 5:

  Input: "-91283472332"
  Output: -2147483648

  Explanation: The number "-91283472332" is out of the range of a 32-bit signed
  integer. Therefore INT_MIN (−2^31) is returned.
*/

#include <string>
#include <iostream>

using namespace std;

class Solution {
public:
    int myAtoi(string str) {
        if (str.empty()) {
            return 0;
        }
        long result = 0;
        bool isNegative = false;

        // 1. Deal with the forward whitespace
        string::size_type i;
        for (i = 0; i < str.length(); i++) {
            if (str[i] != ' ') {
                break;
            }
        }

        // 2. Deal with the sign bit
        if (str[i] == '+' || str[i] == '-') {
            isNegative = (str[i] == '-');
            i++;
        }

        // 3. Deal with the main part
        for (; i < str.length(); i++) {
            if (str[i] < '0' || str[i] > '9') {
                break;
            }
            result = result * 10 + (str[i] - '0');
            if (result > INT32_MAX || result < INT32_MIN) {
                break;
            }
        }

        if (isNegative) {
            result = -result;
        }
        return (result > INT32_MAX ? INT32_MAX :
                (result < INT32_MIN ? INT32_MIN : result));
    }
};


int main(int argc, char **args) {

    Solution solution = Solution();

    string inputs[] = {"42", "   -42", "4193 with words", "words and 987", "-91283472332"};
    int outputs[5];

    for (int i = 0; i < sizeof(inputs) / sizeof(inputs[0]); i++) {
        outputs[i] = solution.myAtoi(inputs[4]);
    }

    return 0;
}